warp10/src/main/java/io/warp10/standalone/StandaloneDirectoryClient.java

```diff
@@ -680,19 +681,22 @@ public void register(Metadata metadata) throws IOException {
     }
   }
 
-  public synchronized void unregister(Metadata metadata) {
-    if (!classids.containsKey(metadata.getName())) {
-      return;
-    }
+  public void unregister(Metadata metadata) {
     // 128BITS
     long labelsId = GTSHelper.labelsId(this.labelsLongs, metadata.getLabels());
-    if (!metadatas.get(metadata.getName()).containsKey(labelsId)) {
-      return;
-    }
-    metadatas.get(metadata.getName()).remove(labelsId);
-    if (metadatas.get(metadata.getName()).isEmpty()) {
-      metadatas.remove(metadata.getName());
-      classids.remove(metadata.getName());
+
+    synchronized (metadatas) {
+      if (!classids.containsKey(metadata.getName())) {
+        return;
+      }
+      if (!metadatas.get(metadata.getName()).containsKey(labelsId)) {
+        return;
+      }
+      metadatas.get(metadata.getName()).remove(labelsId);
+      if (metadatas.get(metadata.getName()).isEmpty()) {
+        metadatas.remove(metadata.getName());
+        classids.remove(metadata.getName());
```

Comment:
On line 632 (in register), the call to metadatas.get() could throw an NPE if the entry for that class name is removed on line 598. The solution would be to protect the part in register with a syncrhonized block on metadatas

Reply:
So register and unregister might have to be mutually exclusive, hence maybe both synchronized

Reply:
I tried to reduce at the minimum synchronization, avoiding making method synchronized.